QUIC proxy peering

espadolini opened the pull request:
This PR adds experimental support for a new QUIC-based transport for proxy peering. Support is enabled by setting the `TELEPORT_UNSTABLE_QUIC_PROXY_PEERING` envvar to `yes` and proxies that opt into the experimental feature will advertise their support by heartbeating with the `teleport.internal/proxy-peer-quic` label set to `yes`, and will exclusively use the QUIC transport to connect through proxies that carry the same label.

Proxies using the QUIC transport for proxy peering expect to be able to bind a UDP socket on the `peer_listen_addr` address and expect to be able to send UDP packets to other QUIC-enabled proxies on their `peer_public_addr` and to receive packets sent to their own `peer_public_addr`. Enabling or disabling the QUIC transport for an existing proxy (in the host ID sense) is unsupported and will lead to very confusing behavior, and the same can be said for restarting a proxy using the QUIC transport; the feature is strictly for environments where new proxies are rolled out (like a Kubernetes deployment).

The protocol and mode of operation is currently described in the package doc in `lib/proxy/peer/quic/quic.go`.

espadolini commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,357 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		case <-c.Context().Done():
+			return
+		default:
+			log.WarnContext(c.Context(), "0-RTT request is missing nonce, requiring full handshake.")
```
A client implementation that uses connection pooling doesn't have much need for 0-RTT, so the way that the server is written allows for full handshook(?) connections to deliberately pretend that the replay protection doesn't exist.

Does that make sense? Should we just always require a timestamp and a nonce no matter what, and treat missing values as an error?

fspmarshall replied:
That makes sense.  If there is a legitimate usecase for not including the nonce, thats fine I think.  Though if it is an intended usecase, I think a warn log might not be warranted, and we probably want the docs on the `nonce` field to explicitly call out that `nonce` is optional and that omission disables 0rtt behavior.

espadolini replied:
I made it so that at least for this "v1alpha" version of the API we're just going to require replay protection no matter what, it makes the server's job much simpler.

espadolini commented on lib/proxy/peer/quicclient.go:
```diff
@@ -0,0 +1,414 @@
+
+	detach := context.AfterFunc(c.runCtx, func() { _ = sc.Close() })
+	// conn.Context() is canceled when the connection is closed
+	c.wg.Add(1)
```
We're holding a count in the waitgroup from earlier in the function, so we don't need to check the mutex again here.

fspmarshall replied:
Ah, missed that.  Maybe call out why the add is safe in a comment?  WG safety gets got wrong a lot (as I so helpfully demonstrate above), don't want someone making code changes later on to misunderstand this pattern and introduce a potential race/panic.

espadolini replied:
Added an explanation.

fspmarshall commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+
+// Sent from the server to the client as a response to a DialRequest. The
+// message is likewise sent in protobuf binary format, prefixed by its length
+// encoded as a little endian uint32.
```
nit: conventionally data sent over the wire is big endian.  GRPC performs length prefixing using big endian uint32s.  unless there is some compelling reason to not to, I'd suggest sticking with that convention.

espadolini replied:
Counterpoint: it's 2024.

gRPC over HTTP/2 uses big endian for length prefixes because the HTTP/2 spec uses big endian and that's just how they happened to write the spec; protobuf itself uses little endian for any fixed-length integers, so "convention" should clearly not be a factor in any new protocol.

fspmarshall replied:
Alright, ignoring convention...  little endian byte order is an affront to god and nature and has no place in a civilized codebase.  Especially for the case of a home-brewed API that we might be called upon to debug at some point, since visually parsing little endian data is annoying/weird.

codingllama replied:
I also have a slight preference for big endian for over-the-wire data, if nothing else because I would expect it to be the case. That said, as long as it's well documented I'm OK with it.

codingllama reviewed (commented):
Quick (no pun intended) look at API and docs.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -39,10 +52,10 @@ type QUICServerConfig struct {
+	// and TLS 1.3 ciphersuites can't be configured in crypto/tls, but for
+	// consistency's sake this should be passed along from the agent
+	// configuration.
 	CipherSuites []uint16
```
If this has no effect then what is the point in having it?

espadolini replied:
I thought it was the easiest way to prevent "Why aren't you using `utils.TLSConfig`?" questions.

codingllama replied:
Kill the field and move the comment to type itself, or next to the appropriate initialization?

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	tlsConfig     *tls.Config
+	quicConfig    *quic.Config
+
+	mu     sync.Mutex
```
What is mu meant to guard? Please document.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+
+	// runCtx is a context that gets canceled when all connections should be
+	// ungracefully terminated.
+	runCtx context.Context
```
Do you really need a context or is this a glorified channel? Bear in mind the usual advice of not having context fields.

Same for serveCtx.

- https://go.dev/wiki/CodeReviewComments#contexts

espadolini replied:
We make use of `context.AfterFunc` for both.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		MaxStreamReceiveWindow:     quicMaxReceiveWindow,
+		MaxConnectionReceiveWindow: quicMaxReceiveWindow,
+
+		MaxIncomingStreams:    1 << 60,
```
Is there a specific reason for `1<<60` or is it just a large number? Add a comment?

```suggestion
		MaxIncomingStreams:    1 << 60, // arbitrary, large number
```

espadolini replied:
It's a quic-go magic number with no matching const, the maximum possible value for `MaxIncomingStreams` 😢

I'll add a comment pointing that out.

espadolini replied:
const'd in 5fad0d54d9

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+			return trace.Wrap(err)
+		}
+
+		s.wg.Add(1)
+		go s.handleConn(c)
```
Don't spread out concurrency control, that makes ownership loose and makes it harder to reason about.

```suggestion
		s.wg.Add(1)
		go func() {
			s.handleConn(c)
			s.wg.Done()
		}()
```

espadolini replied:
That would make sense if not for the fact that `handleConn` still needs to interact with the waitgroup when spawning goroutines for each stream - and this is actually somewhat important, because moving the `wg.Done()` out of `handleConn` would mean that the `wg.Add(1)` in it becomes very suspect, as it wouldn't be "guarded" by an active waitgroup count and thus could potentially race against waitgroup waiters.

codingllama replied:
I politely disagree - both Serve and handleConn could use this same pattern to take care of the goroutines they spawn. A method shouldn't have to know it's running as a goroutine, or mandate this kind of setup to be called.

espadolini replied:
Added a per-connection waitgroup that counts streams and is waited on before returning from the connection handler.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	}
+}
+
+func (s *QUICServer) handleConn(c quic.EarlyConnection) {
```
nit: Pass the context down from Serve?

```suggestion
func (s *QUICServer) handleConn(ctx context.Context, conn quic.EarlyConnection) {
```

espadolini replied:
Nothing but the `AfterFunc(runCtx)` in there should be using a context other than the connection's context anyway - I'll update the function to do that.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
 }
 
 // QUICServer is a proxy peering server that uses the QUIC protocol.
```
That nice QUIC peering overview you wrote could go here, for example.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		"internal_id", uuid.NewString(),
+	)
+	state := c.ConnectionState()
+	log.InfoContext(s.serveCtx,
```
Start is logged is logged using s.serveCtx, connection closed is logged using c.Context() (line 220), down below we do AcceptStream using context.Background(). I'm not sure I follow the logic, shouldn't all of them either use s.serveCtx or c.Context() ?

espadolini replied:
(see answer to https://github.com/gravitational/teleport/pull/47587#discussion_r1803818739)

espadolini replied:
(`AcceptStream` should run until the connection is closed, anyway - we don't currently have a way to gracefully reject streams, I'd like for quic-go to get the ability to alter the MaxIncomingStream value on a running connection first, so we can set that to 0 and then consume all the queued streams)

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+			return
+		}
+
+		s.wg.Add(1)
+		go s.handleStream(st, c, log)
```
Same for spreading out concurrency control:

```suggestion
		s.wg.Add(1)
		go func() {
			s.handleStream(st, c, log)
			s.wg.Done()
		}()
```

espadolini replied:
This one I could see having a point - but everything is slightly tied together anyway, is there really a point?

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	}
+}
+
+func (s *QUICServer) handleStream(st quic.Stream, c quic.EarlyConnection, log *slog.Logger) {
```
nit:

```suggestion
func (s *QUICServer) handleStream(
  ctx context.Context,
  stream quic.Stream,
  conn quic.EarlyConnection,
  log *slog.Logger,
) {
```

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	defer st.CancelRead(0)
+	defer st.CancelWrite(0)
+
+	log.DebugContext(c.Context(), "Handling stream.")
```
nit: Do this before the "done handling stream" log

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	defer s.wg.Done()
+
+	log = log.With("stream_id", st.StreamID())
+	defer log.DebugContext(c.Context(), "Done handling stream.")
+
+	defer st.CancelRead(0)
+	defer st.CancelWrite(0)
```
Maybe combine the defers?

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		if err != nil {
+			return
+		}
+		st.SetWriteDeadline(time.Now().Add(10 * time.Second))
```
Use a const for the deadline?

Why 10s? Please document.

Same below.

espadolini replied:
Arbitrary magic number (with the only constraint that it should probably result in the server giving up on the dial earlier than our client), I'll move it together with the other constants.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+			return
+		}
+		st.SetWriteDeadline(time.Now().Add(10 * time.Second))
+		if _, err := st.Write(binary.LittleEndian.AppendUint32(nil, uint32(len(errBuf)))); err != nil {
```
I know it's a bit... exaggerated... but should we make sure the message fits in an uint32?

espadolini replied:
We absolutely should.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	s.closed = true
+	s.mu.Unlock()
+
+	defer s.runCancel()
+	defer context.AfterFunc(ctx, s.runCancel)()
+	s.serveCancel()
+	s.wg.Wait()
```
```suggestion
	defer context.AfterFunc(ctx, s.runCancel)()
	s.serveCancel()
	s.wg.Wait()
	s.runCancel()
```

espadolini replied:
I've been liking the pattern of
```go
defer idempotentFoo()
defer context.AfterFunc(ctx, idempotentFoo)()
```

to cancel the `AfterFunc` because we're obviously going to run the callback ourselves - which also guarantees that we're not leaving the callback attached to the context.

codingllama replied:
It's a subtlety that might be lost in most people looking at this. Works fine so I'll resolve.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+// needed.
+type replayStore struct {
+	mu   sync.Mutex
+	t    time.Time
```
nit: Use a more descriptive name for `t`.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	mu   sync.Mutex
+	t    time.Time
+	cur  map[uint64]struct{}
+	prev map[uint64]struct{}
```
Why do we keep 2 time windows?

espadolini replied:
Explained a bit more.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	s.closed = true
+	s.mu.Unlock()
+
+	defer s.runCancel()
+	defer context.AfterFunc(ctx, s.runCancel)()
```
I question this a bit, if the caller wants non-graceful shutdown then shouldn't they call Close?

espadolini replied:
The various `Shutdown` functions I've seen (in our codebase and outside) seem to use the context as an indicator of a grace period, with the context being canceled triggering an ungraceful close. Is that not expected? I'm not sure if it's ever particularly meaningful to partially abort a shutdown.

espadolini replied:
This is roughly the behavior expected by `proxy.shutdown` so I'll stick with this - I'll be amenable to choosing a different name but I can't think of one. The alternative is calling shutdown and then unconditionally calling close (or conditionally calling close if the error is a context cancellation), and seeing as we'd be doing that in 100% (1/1) of the callsites for the method, it makes sense to have it be part of the method itself.

codingllama replied:
I disagree about it making sense to be part of the method itself, as methods tend to be better served by having a single, clear purpose. I'll yield on the grounds of "proxy.shutdown is already weird", though.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+
+	log.DebugContext(c.Context(), "Handling stream.")
+
+	sendErr := func(toSend error) {
```
Is "sendErr" any different from a regular send? Maybe have a `QUICServer.send(context.Context, proto.Message)` method instead?

espadolini replied:
Currently we send an error trail-converted to grpc in the status field of the message or we send nothing; if we need something generic in the future we'll make it generic.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		}
+	}
+
+	st.SetReadDeadline(time.Now().Add(10 * time.Second))
```
Refactor receive logic in its own method?

codingllama commented on lib/proxy/peer/quic.go:
```diff
@@ -0,0 +1,141 @@
+)
+
+const (
+	quicMaxIdleTimeout  = 30 * time.Second
+	quicKeepAlivePeriod = 5 * time.Second
```
Are these arbitrary? Please document.

espadolini replied:
They are!

codingllama commented on lib/proxy/peer/quic.go:
```diff
@@ -0,0 +1,141 @@
+	// quicMaxMessageSize is the maximum accepted size (in protobuf binary
+	// format) for the request and response messages exchanged as part of the
+	// dialing.
+	quicMaxMessageSize = 128 * 1024
```
This is much smaller than the uint32 we keep mentioning. Why is that?

espadolini replied:
Seems like a reasonable size, and if we tuned it down to 65535 and used a uint16 we wouldn't have any way to make it bigger without changing the protocol, if it came down to it.

codingllama replied:
Got it. Let's mention that this is somewhat arbitrary and that the reason for the uint32 is that we want room to grow.

Should we limit the messages sent in the other direction to that size too?

espadolini replied:
Both directions are limited now, we log a message with the ever helpful "(this is a bug)" instead of sending an oversized message.

codingllama replied:
If we hit the bug and abort the send, what happens to the connection? Should we make an effort to truncate?

codingllama commented on lib/proxy/peer/quic.go:
```diff
@@ -0,0 +1,141 @@
+
+	// quicNextProto is the ALPN indicator for the current version of the QUIC
+	// proxy peering protocol.
+	quicNextProto = "teleport-peer-v1a"
```
```suggestion
	quicNextProto = "teleport-peer-v1alpha1"
```

espadolini replied:
Sure - although we should really think about shrinking our nextprotos rather than lengthen them; too big of a clienthello can be problematic 😬

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		return
+	}
+	if reqLen >= quicMaxMessageSize {
+		log.WarnContext(c.Context(), "Received oversized request.", "request_len", reqLen)
```
Is this warning in any way actionable for someone reading the logs?

espadolini replied:
It would show the misbehaving peer as part of the log fields, but it's either a malicious client or a bug.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		select {
+		case <-c.HandshakeComplete():
+		case <-c.Context().Done():
+			return
```
Log the timeout?

espadolini replied:
I'm always weary of logging in places where we're likely to just get swarmed with "EOF" "context canceled" " connection reset by peer", although maybe here specifically it could be useful to diagnose weirdness.

codingllama replied:
We can use debug or trace level if you think it's appropriate, but I'd rather have the toggle and not use it than the alternative.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	// a replayed request is always wrong even after a full handshake, the
+	// replay might've happened before the legitimate request
+	if !s.replayStore.add(req.GetNonce(), time.Now()) {
+		log.ErrorContext(c.Context(), "Request is reusing a nonce, rejecting.", "nonce", req.GetNonce())
```
Same for the log being actionable.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		return
+	}
+
+	_, clusterName, ok := strings.Cut(req.GetServerId(), ".")
```
Please document the expected format of server_id in the proto.

espadolini replied:
Logic copied wholesale from gRPC proxy peering.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		// an empty protobuf message has an empty wire encoding, so by sending a
+		// size of 0 (i.e. four zero bytes) we are sending an empty DialResponse
+		// with an empty Status, which signifies a successful dial
+		if _, err := st.Write(binary.LittleEndian.AppendUint32(nil, 0)); err != nil {
```
This is clever and all but I'd rather we just did a send() call with the corresponding "OK" message.

espadolini replied:
I don't want to bring in a fallible operation that allocates a bunch of times, initializes hidden state and is fallible just to send four NULs.

codingllama replied:
You can marshal an empty message once and reuse if you like, but the readability is needlessly suffering here.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		_, err := io.Copy(nodeConn, st)
+		return trace.Wrap(err)
+	})
+	_ = eg.Wait()
```
Log the error?

espadolini replied:
I'll give it a shot, I suspect it would be pretty spammy tho.

codingllama replied:
Same for debug or trace level.

espadolini replied:
Errors with a zero code should now be appropriately ignored in log messages; stream and connection closing before the connection is established (or successfully fails(???)) now use a nonzero error code which is still quite aspecific but we can figure that out at a later time.

codingllama commented on lib/proxy/peer/quicserver.go:
```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		// available streams during a connection (so we can set it to 0)
+		st, err := c.AcceptStream(context.Background())
+		if err != nil {
+			log.DebugContext(c.Context(), "Got an error accepting a stream.", "error", err)
```
As a general comment, I'd rather this function returned an error and the caller made the choice to swallow it.

espadolini replied:
The only possible error is caused by the connection getting closed - which is also why the log line is at debug level. I'm not convinced that moving the error logging one layer above will do much for the clarity of the code, at least while this is the only exit point for the function.

codingllama replied:
Moving the logging up makes this behave like a regular erroring function, which is already a valuable readability improvement IMO.

espadolini replied:
The logging happens before the defers, logging after returning would mean that the log line is related to an error that happened potentially much earlier, seeing as now we're waiting for the per-connection waitgroup to end.

codingllama replied:
That is not typically an issue - a function errors, runs defers, returns the error and then the error gets logged by the caller. I won't push further but I do think a regular erroring func tends to be simpler to follow than the unusual over-swallowing of errors.

codingllama commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+
+syntax = "proto3";
+
+package teleport.quicpeering.v1alpha;
```
Maybe:

```suggestion
package teleport.quicpeering.v1alpha1;
```

Bold of you to assume there will be only one alpha... ;)

codingllama replied:
I guess it really depends on whether we expect more alphas. If yes, then go v1alpha1. Otherwise we can do v1alpha to v1alpha2 like you said.

codingllama reviewed (commented):
Server review. I think you would benefit from someone who understands QUIC (if we have any), but I did my best.

I'll wait for replies before catching up to the client, so you have time to catch up to it all.

espadolini commented on lib/proxy/peer/client.go:
```diff
@@ -72,7 +72,7 @@ type ClientConfig struct {
 	TLSCipherSuites []uint16
 	// Log is the proxy client logger.
-	Log logrus.FieldLogger
+	Log *slog.Logger
```
Stacked this PR on top of #47649; sorry about the rebase and the slight resulting tweaks to commits.

espadolini commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+// binary format, prefixed by its length encoded as a little endian uint32.
+message DialRequest {
+  string server_id = 1;
+  string conn_type = 2;
```
Any connection type that can get registered as a tunnel or that can be interpreted as a valid direct dial, technically (although the latter will always be rejected on the account of being a proxy peering dial). The list of const values for `api/types.TunnelType` should be exhaustive.

espadolini replied:
Explained in f804827f05

espadolini commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+  // server time for 0-RTT requests.
+  google.protobuf.Timestamp timestamp = 5;
+  // A random id for each dial request, must be provided and unique among
+  // requests presented in a 10-20 minute window for 0-RTT requests.
```
That's an old value, it should read "at least 5 minutes". It's not an exact cutoff because having to keep track of exact "expiration" times is pretty silly since we're supposed to have 64 bits of entropy (which is not enough to guarantee uniqueness long-term, but is definitely enough for a much longer time than 5 or 10 minutes' worth of proxy peering connections), so the implementation I chose just swaps between two hashsets at the first request past the 5 minute cutoff time.

espadolini replied:
Fixed in f804827f05

espadolini commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+
+// A stringy Go net.Addr.
+message Addr {
+  string network = 1;
```
No idea if anything but "tcp" is allowed but I would sincerely hope to avoid turning this into an exhaustive analysis of lib/reversetunnel 😩

espadolini replied:
Added examples in f804827f05

espadolini commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+// A stringy Go net.Addr.
+message Addr {
+  string network = 1;
+  string addr = 2;
```
This just matches the "anything goes" `net.Addr` implementation that we are currently using for regular gRPC proxy peering - see comment about `network`.

espadolini replied:
Elaborated a bit in f804827f05

espadolini commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+
+// Sent from the client to the server in a fresh QUIC stream to dial a server
+// through a QUIC proxy peering connection. The message is sent in protobuf
+// binary format, prefixed by its length encoded as a little endian uint32.
```
Nothing good, so I'm going to add a check for that at marshaling time.

espadolini replied:
Added in c32aaa53d9 (albeit just as a check against sending messages bigger than 128KiB, serving a double purpose)

espadolini commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+  string server_id = 1;
+  string conn_type = 2;
+
+  Addr source = 3;
+  Addr destination = 4;
```
The source is the address of the user initiating the connection to the proxy acting as the proxy peering client, which gets propagated to the final target of the connection; the destination is intended to act as a hint for direct dialing that will never take place since this is a proxy peering connection but we're like three levels of vestigiality deep into reverse tunnels here - morally (and practically), it's the `LocalAddr` of the reified `net.Conn` finally handled by the agent at the opposite end of the connection than the user.

espadolini replied:
Explained in f804827f05

espadolini commented on lib/proxy/peer/quic.go:
```diff
@@ -0,0 +1,141 @@
+If the status is ok (signifying no error) then the stream will stay open,
+carrying the data for the connection between the user and the agent, otherwise
+the stream will be closed. For sanity's sake, the size of both messages is
+limited and any oversized message is treated as a protocol violation.
```
We read the size first, if the message is oversized we just close the stream.

codingllama replied:
I thought we capped at uint32, but later I saw we have a limit on top of the uint32, which then made this line make sense. Maybe we should mention that we have an arbitrary limit, so the messages can't use the full uint32 length?

espadolini commented on lib/proxy/peer/quic.go:
```diff
@@ -0,0 +1,141 @@
+request sent as 0-RTT can result in at most one connection opened through an
+agent tunnel.
+
+The client should make sure to not send data belonging to the connection as part
+of the early data; this will cause no further delays if the client intends to
```
I'll elaborate a bit more - early data lets us transmit the dial request immediately, but a very optimistic client could decide to also include some data that should belong to the tunneled connection immediately, before receiving the success response.

This is just an additional insurance policy - if we get replay protection wrong, the worst an attacker can do is create load by opening new connections that time out with no data being sent from the client to the server.

espadolini replied:
Added a tiny bit more explanation in c03013f30c

espadolini commented on lib/proxy/peer/quic.go:
```diff
@@ -0,0 +1,141 @@
+The client should make sure to not send data belonging to the connection as part
+of the early data; this will cause no further delays if the client intends to
+wait for the server to reply to the dial request. A client that wants to make
+use of multiplexing should take care to not accidentally send more than one dial
+request as 0-RTT in a single connection.
```
Otherwise an attacker gets to replay two requests at once, which can cause slightly more unauthenticated load than we absolutely have to handle.

espadolini replied:
Mentioned this in c03013f30c

espadolini commented on lib/proxy/peer/quicclient.go:
```diff
@@ -0,0 +1,441 @@
+}
+
+func (e wrongProxyError) Unwrap() error {
+	return &trace.AccessDeniedError{
```
The `wrongProxyError` type is checked against when deciding if we should log the warning for the cluster admin that there's a misconfiguration in the proxies' public peer addresses - we could check for an `AccessDeniedError` with an exact message I suppose, but why?

codingllama replied:
I didn't read all of quicclient.go, so I missed the type check. I'll resolve.

espadolini commented on lib/service/servicecfg/proxy.go:
```diff
@@ -158,6 +158,11 @@ type ProxyConfig struct {
+	// QUICProxyPeering will make it so that proxy peering will support inbound
+	// QUIC connections and will use QUIC to connect to peer proxies that
+	// advertise support for it.
+	QUICProxyPeering bool
```
`servicecfg.ProxyConfig` is internal use and filled only by code, not by unmarshaling a config file or user provided options, so we can expand on it when we have to.

codingllama replied:
Fair enough.

espadolini commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -0,0 +1,57 @@
+
+option go_package = "github.com/gravitational/teleport/gen/proto/go/teleport/quicpeering/v1alpha;quicpeeringv1alpha";
+
+// Sent from the client to the server in a fresh QUIC stream to dial a server
```
Did so in f804827f05

codingllama commented on proto/teleport/quicpeering/v1alpha/dial.proto:
```diff
@@ -23,27 +23,37 @@ import "google/rpc/status.proto";
+// Sent from a proxy to a peer proxy in a fresh QUIC stream to dial a Teleport
+// resource through a QUIC proxy peering connection. The message is sent in
+// protobuf binary format, prefixed by its length encoded as a little endian
+// 32-bit integer.
```
```suggestion
// 32-bit unsigned integer.
```

codingllama commented on lib/service/service.go:
```diff
@@ -4767,13 +4771,13 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 			process.RegisterCriticalFunc("proxy.peer.quic", func() error {
 				if _, err := process.WaitForEvent(process.ExitContext(), ProxyReverseTunnelReady); err != nil {
-					logger.DebugContext(process.ExitContext(), "Process exiting: failed to start QUIC peer proxy service waiting for reverse tunnel server.")
+					logger.DebugContext(process.ExitContext(), "process exiting: failed to start QUIC peer proxy service waiting for reverse tunnel server")
```
nit: I think the "agreed" logging style is to capitalize the first word but avoid punctuation.

See the examples in RFD 0154 and this section:

> The message should be a fragment, and not a full sentence. Terminating the message with punctuation should be avoided.

* https://github.com/gravitational/teleport/blob/master/rfd/0154-logging-guidelines.md#logging-standards

espadolini replied:
Word of @rosstimothy is that the message shouldn't be capitalized and "the message should just be another key-value pair" (presumably with "message" as the key), although I agree that the RFD could be clearer on that.

codingllama replied:
We really should update the examples and add a section stating the preferred style, I think very few of us are getting this "right" (me included apparently).

codingllama reviewed (commented):
Thanks for the added documentation/context. Very helpful!

espadolini commented on lib/proxy/peer/quic.go:
```diff
@@ -0,0 +1,141 @@
+
+/*
+
+# QUIC proxy peering
```
Moved packages around, now it's the godoc for the inner `lib/proxy/peer/quic` package.

espadolini commented:
@codingllama after shuffling some types and functions around, QUIC proxy peering is now in its own package

espadolini commented on lib/proxy/peer/quic/client.go:
```diff
@@ -0,0 +1,553 @@
+// Close implements [internal.ClientConn].
+func (c *quicClientConn) Close() error {
+	c.mu.Lock()
+	c.closed = true
```
`Close` and `Shutdown` are both idempotent, so it's ok, we don't need to check for it (and I don't see a lot of value in returning some `ErrClosed` on a double `Close`).

rosstimothy reviewed (approved):
Ran through a few tests on this branch again and noticed one last log spam nit. I was seeing the following in the proxy logs at the conclusion of every tsh ssh session:

```
2024-11-07T11:56:47-05:00 DEBU [PROXY:QPE] done forwarding data pid:18690.1 remote_addr:127.0.0.1:4021 internal_id:401212a6-d467-487d-8424-06bbe1bdf215 stream_id:0 error:[
ERROR REPORT:
Original Error: *trace.ConnectionProblemError use of closed network connection
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/utils/sshutils/chconn.go:141 github.com/gravitational/teleport/api/utils/sshutils.(*ChConn).Read
	github.com/gravitational/teleport/lib/reversetunnel/conn_metric.go:80 github.com/gravitational/teleport/lib/reversetunnel.(*metricConn).Read
	io/io.go:429 io.copyBuffer
	io/io.go:388 io.Copy
	github.com/gravitational/teleport/lib/proxy/peer/quic/server.go:389 github.com/gravitational/teleport/lib/proxy/peer/quic.(*Server).handleStream.func3
	golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78 golang.org/x/sync/errgroup.(*Group).Go.func1
	runtime/asm_arm64.s:1223 runtime.goexit
User Message: use of closed network connection] quic/server.go:415
```

codingllama commented on lib/proxy/peer/quic/quic.go:
```diff
@@ -0,0 +1,189 @@
+		return nil, trace.Wrap(err)
+	}
+	if len(buf)-4 > math.MaxUint32 {
+		return nil, trace.LimitExceeded("oversized message")
```
Is trace.LimitExceeded correct? Isn't that more related to rate-limiting than an message-size failure?

espadolini replied:
`LimitExceeded` maps to `codes.ResourceExhausted`, which is generated when `grpc-go` deals with an oversized message. Rate limiting would probably be `codes.Unavailable` (which weirdly enough maps to `ConnectionProblemError` in `trace/trail`).

codingllama commented on lib/proxy/peer/quic/quic.go:
```diff
@@ -0,0 +1,189 @@
+// marshalSized returns the wire encoding of the given [proto.Message] prefixed
+// by its length encoded as a little endian 32-bit integer.
+func marshalSized(m proto.Message) ([]byte, error) {
+	buf := make([]byte, 4, 4+proto.MarshalOptions{}.Size(m))
```
Add a brief comment explaining why length=4 capacity=4+size?

codingllama commented on lib/proxy/peer/quic/quic.go:
```diff
@@ -0,0 +1,189 @@
+	if appErr := (*quic.ApplicationError)(nil); errors.As(err, &appErr) && appErr.ErrorCode == 0 {
+		return nil
+	}
+	return trace.Wrap(err)
```
I don't think the trace.Wrap belongs here, let the caller wrap if they want to. It's documented, which is good, but it's still surprising.

codingllama reviewed (commented):
Apologies for the delay. Here's a pass on the "easy" parts, I'm still giving the server/client parts a proper look.

I probably said this before, but a non-trivial 2.5k lines, 45-commits PR is pretty difficult to hold in ones head/reason about. This would benefit from being split into human-friendly sized PRs.

espadolini commented on lib/proxy/peer/internal/tls.go:
```diff
@@ -0,0 +1,85 @@
+	return nil
+}
+
+func VerifyPeerCertificateIsSpecificProxy(peerID string) verifyPeerCertificateFunc {
```
`verifyPeerCertificateFunc` is not a private type, it's an alias for the func type used in `tls.Config.VerifyPeerCertificate` - I'll add a comment to point that out.

codingllama replied:
The alias being private makes a difference. Maybe make the alias itself public, plus the comment?

espadolini replied:
Removed the alias altogether.

codingllama commented on lib/proxy/peer/internal/tls.go:
```diff
@@ -0,0 +1,85 @@
+
+const DuplicatePeerMsg = "Detected multiple Proxy Peers with the same public address when connecting to a Proxy which can lead to inconsistent state and problems establishing sessions. For best results ensure that `peer_public_addr` is unique per proxy and not a load balancer."
+
+type WrongProxyError struct{}
```
It's not plain string matching, it takes the AccessDeniedError type into account as well. It also matches your impl, as an errors.Is with that exact same instance still matches WrongProxyError because of your Unwrap.

espadolini replied:
It will cease to match if we change the message somewhere but not somewhere else, and it will also match against a completely unrelated `trace.AccessDenied("connected to unexpected proxy")`.

codingllama replied:
I did mean for you to keep it as a global var, and for other places to keep using it. I'm just not seeing what benefits the type brings.

This already matches an unrelated `trace.AccessDenied("connected to unexpected proxy")` error. It's worth asking if that's really a problem, or if they are indeed distinct if they match so exactly.

espadolini replied:
Satisfying `trace.IsAccessDenied` is a property that I didn't want to break from the old code; the match is only in one direction, not the other:
```go
require.ErrorIs(t, WrongProxyError{}, trace.AccessDenied(WrongProxyError{}.Error()))
require.NotErrorIs(t, trace.AccessDenied(WrongProxyError{}.Error()), WrongProxyError{})
```

i.e. a `WrongProxyError` is a `*trace.AccessDeniedError`, but not viceversa.

codingllama replied:
I still think this could be quite a ways simpler, but I'll leave it to you.

codingllama commented on lib/reversetunnelclient/peer.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+// NewPeerDialer implements [peerdial.Dialer] for a reverse tunnel server.
+func NewPeerDialer(server Server) PeerDialerFunc {
```
Feel free to push changes to a follow-up if you think it's best.

espadolini replied:
Changing this to use the existing `Tunnel` interface, which is `GetSite` and `GetSites`.

espadolini commented on lib/proxy/peer/dial/dial.go:
```diff
@@ -0,0 +1,38 @@
+	To       *utils.NetAddr
+	ServerID string
+	ConnType types.TunnelType
+}
```
Doesn't work because of package import cycles, even getting rid of the `reversetunnelclient.Server` in param position you're still going to have a `GetSite` that returns an object with a `Dial` method that takes a `reversetunnelclient.DialParams` - and requiring the caller to do three layers of rewrapping is almost the entire point of this function, afaict.

codingllama replied:
Thanks for the explanation.

espadolini commented on lib/proxy/peer/internal/clientconn.go:
```diff
@@ -0,0 +1,50 @@
+	// immediately.
+	Close() error
+
+	// Shutdown waits until all connections are closed or the context is done,
```
More like "will prevent the opening of new connection" (this is the client side), but that's not even true for `grpcClientConn`, which will happily allow the caller to use the connection while it's still open and running other connections. 🥲

espadolini commented on lib/proxy/peer/client.go:
```diff
@@ -205,13 +181,13 @@ type grpcClientConn struct {
 
-// peerID implements [clientConn].
-func (c *grpcClientConn) peerID() string { return c.id }
+// PeerID implements [internal.ClientConn].
```
I'm not sure about that, it signals (and links, with godoc links) which interface we were intending to implement with this method.

espadolini commented on lib/proxy/peer/internal/tls.go:
```diff
@@ -0,0 +1,85 @@
+
+type verifyPeerCertificateFunc = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
+
+var _ verifyPeerCertificateFunc = VerifyPeerCertificateIsProxy
```
Removed the alias here.

espadolini committed:
Docs, utilities and protobuf definitions

espadolini committed:
Server implementation

espadolini committed:
Client implementation

espadolini committed:
Tests

espadolini commented:
@codingllama I've split off the code reorganization parts in #48836, but I don't really see a sensible way to split off the client and server parts of the QUIC implementation without it being somewhat meaningless. I've rebased the commits tho, they should be far more manageable now.

codingllama commented on lib/proxy/peer/quic/server.go:
```diff
@@ -70,31 +84,395 @@ func (c *ServerConfig) checkAndSetDefaults() error {
+		return
+	}
+
+	var eg errgroup.Group
```
Set the group limit to 2, to be explicit/safe.

Should we use the group context to cancel in-flight goroutines, or is that redundant? This probably deserves a brief comment.

espadolini replied:
We call Go twice in succession and then immediately call Wait, we don't move or reuse the errgroup, and there's no errgroup context, I'm not entirely sure what the comment should be.

codingllama commented on lib/proxy/peer/quic/server.go:
```diff
@@ -70,31 +84,395 @@ func (c *ServerConfig) checkAndSetDefaults() error {
 }
 
+// dialResponseOK is the length-prefixed encoding of a DialResponse message that
+// signifies a successful dial.
```
```suggestion
// signifies a successful dial.
// See TestDialResponseOKEncoding.
```

espadolini committed:
Log the correct handshake error

espadolini commented on lib/proxy/peer/quic/client.go:
```diff
@@ -0,0 +1,543 @@
+
+// Read implements [net.Conn].
+func (c *streamConn) Read(b []byte) (n int, err error) {
+	return c.st.Read(b)
```
`quic.Stream` has methods that I don't want to expose, unfortunately.

codingllama replied:
Guarding the interfaces, I respect that.

espadolini commented on lib/proxy/peer/quic/client.go:
```diff
@@ -0,0 +1,543 @@
+				"failed to complete handshake after exchanging 0-RTT dial request",
+				"error", err,
+			)
+			return nil, trace.Wrap(context.Cause(earlyConn.Context()))
```
Yes! `quic-go` sets the error that caused the connection to close as the cancellation cause of the connection context.

codingllama replied:
> Yes! quic-go sets the error (...)

That's nice! So far I've seen little reason to worry about/consider cancel causes.

espadolini committed:
Use constants for stream and application error codes

espadolini committed:
Avoid stuttering in const names

espadolini commented on lib/proxy/peer/quic/client.go:
```diff
@@ -0,0 +1,543 @@
+		dst: dst,
+	}
+
+	detach := context.AfterFunc(c.runCtx, func() { _ = sc.Close() })
```
Not really, if we hit that it's because the whole `ClientConn` has been closed, logging will just add a wall of noise with no particular benefit.

espadolini commented on lib/proxy/peer/quic/client.go:
```diff
@@ -0,0 +1,543 @@
+		return trace.Wrap(err)
+	}
+	stream.CancelRead(0)
+	_ = stream.Close()
```
I don't think we care about the stream error if we successfully exchanged the ping request and status response.

espadolini commented on lib/proxy/peer/quic/quic_test.go:
```diff
@@ -0,0 +1,414 @@
+		types.NodeTunnel,
+	)
+	if err == nil {
+		_ = conn.Close()
```
It's not supposed to happen, but it doesn't feel great to leak an `io.Closer` just because the testing logic has decided to terminate the test early. Replaced the check with an `assert.Error` instead.

codingllama replied:
Yep, a t.Error or something like that helps a lot in debugging if/when the impossible happens.

espadolini commented on lib/proxy/peer/quic/quic_test.go:
```diff
@@ -0,0 +1,414 @@
+	})
+	require.NoError(t, err)
+	go server.Serve(serverTransport)
+	t.Cleanup(func() { _ = server.Close() })
```
Connection closing errors are not particularly helpful, the logic tests for successful data transfer, which is what the user of the connection will care about.

espadolini committed:
Improve godocs and explanations

espadolini commented on lib/proxy/peer/quic/quic_test.go:
```diff
@@ -0,0 +1,414 @@
+			clientCert: clientProxyCert,
+			rootCAs:    correctCAPool,
+			check: func(t require.TestingT, err error, msgAndArgs ...interface{}) {
+				require.Error(t, err, msgAndArgs...)
```
It's clearer to have an explicit "expected an error, got nil".

espadolini commented on lib/proxy/peer/quic/quic_test.go:
```diff
@@ -0,0 +1,414 @@
+				var transportError *quic.TransportError
+				require.ErrorAs(t, err, &transportError, msgAndArgs...)
+				require.True(t, transportError.Remote, msgAndArgs...)
+				const quicAlertUnknownCA quic.TransportErrorCode = 48 + 0x100
```
Not sure if it's a QUIC thing or a `quic-go` thing, but TLS alerts are surfaced with a transport error code of 0x100 plus the alert code, which comes from the TLS RFC (and/or some unexported consts in `crypto/tls`). Added a comment to explain both things.

codingllama replied:
Much better, thanks.

espadolini commented on lib/proxy/peer/quic/quic_test.go:
```diff
@@ -0,0 +1,414 @@
+
+	require.NoError(t, conn.Close())
+	t.Log("closed")
+	<-pipeClose
```
Would it be a clean failure, or a timeout that doesn't respect the test timeout as specified by the harness?

codingllama replied:
The global test timeout is usually too high, so I think it's worth adding an explicit 1s-2s here in case it locks for some reason. (As of now obviously this works, so it would have little effect.)

espadolini committed:
Address test review comments
